### ISSUE TITLE
stability: lease and freeze handling for proposer-evaluated KV

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#10414]
At the time of writing, proposer-evaluated KV does not correctly deal with leases. More precisely, the logic downstream of Raft is missing:

* it is not verified that a command applies under the lease it was proposed (which is necessary)
* `{Request,Transfer}Lease` have no special downstream-of-Raft handling which is necessary to avoid lease clobbering.

Similarly, `ChangeFrozen` (or any other command that can be proposed by anyone) needs such special handling. The changes should be made in a way that makes sure that we can't accidentally change the semantics of existing commands or add new ones that would also need special casing.

Assigning @petermattis for proper reassignment.

**Comment by spencerkimball** [comment]
@bdarnell could you elaborate on what @tschottdorf meant by "avoiding lease clobbering" downstream-of-Raft? I don't understand the scenario from the description.

By the way, are we currently hoping to eliminate the freeze cluster machinery?

**Comment by bdarnell** [comment]
The problem is that the RequestLease method bypasses most of the places where we validate leases (it has the skipLease flag set). The only place where we validate the proposed lease is in `Replica.RequestLease` itself. Currently, that's called downstream of raft where it has its place in the log so we know it's up to date. In propEvalKV, it's called upstream of raft, so its information might be stale and it will be allowed into the raft log at any point. We need to validate the lease in processRaftCommand whether the skipLease flag is present or not (skipLease would only be used in upstream-of-raft checks). But since the propEvalKV issue was there previously, I'd rather leave that for a separate PR after #10305 goes in.

As for freeze, I think we'll be able to get rid of it. I was planning to keep it around because it's not hurting anything, but this might be a reason to drop it sooner rather than later. Let's keep it around for now (until I'm more confident that the propEvalKV switch can happen without it), but don't spend any time updating it for this issue.
